plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestReusingConnectionFactory.java

```diff
@@ -0,0 +1,166 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.security.ConnectorIdentity;
+import io.trino.testing.TestingConnectorSession;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Duration;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Thread.sleep;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class TestReusingConnectionFactory
+{
+    private static final ConnectorSession ALICE = TestingConnectorSession.builder()
+            .setIdentity(ConnectorIdentity.ofUser("alice"))
+            .build();
+    private static final ConnectorSession BOB = TestingConnectorSession.builder()
+            .setIdentity(ConnectorIdentity.ofUser("bob"))
+            .build();
+
+    @Test
+    public void testConnectionIsReusedSingleUser()
+            throws Exception
+    {
+        try (MockConnectionFactory mockConnectionFactory = new MockConnectionFactory();
+                ReusingConnectionFactory connectionFactory = new ReusingConnectionFactory(mockConnectionFactory, Duration.ofSeconds(5), 10)) {
+            connectionFactory.openConnection(ALICE).close();
+            connectionFactory.openConnection(ALICE).close();
+            assertThat(mockConnectionFactory.openedConnections).isEqualTo(1);
+        }
+    }
+
+    @Test
+    public void testConnectionIsReusedTwoUsers()
+            throws Exception
+    {
+        try (MockConnectionFactory mockConnectionFactory = new MockConnectionFactory();
+                ReusingConnectionFactory connectionFactory = new ReusingConnectionFactory(mockConnectionFactory, Duration.ofSeconds(5), 10)) {
+            connectionFactory.openConnection(ALICE).close();
+            connectionFactory.openConnection(BOB).close();
+            connectionFactory.openConnection(BOB).close();
+            connectionFactory.openConnection(ALICE).close();
+            assertThat(mockConnectionFactory.openedConnections).isEqualTo(2);
+        }
+    }
+
+    @Test
+    public void testConnectionIsNotShared()
+            throws Exception
+    {
+        try (MockConnectionFactory mockConnectionFactory = new MockConnectionFactory();
+                ReusingConnectionFactory connectionFactory = new ReusingConnectionFactory(mockConnectionFactory, Duration.ofSeconds(5), 10);
+                Connection ignored = connectionFactory.openConnection(ALICE)) {
+            connectionFactory.openConnection(ALICE).close();
+            assertThat(mockConnectionFactory.openedConnections).isEqualTo(2);
+        }
```

Comment:
assert that now one is retained in the cache and one underlying has been closed already